### ISSUE TITLE
Document that certain asynchronous methods block inside Routines

### ISSUE THREAD

**Issue by Spacechild1** [supercollider/supercollider#6448]
Asynchronous methods like `Buffer.loadToFloatArray` use `forkIfNeeded` behind the scene, which means that they actually block when executed inside a Routine! Unfortunately, this behavior is currently not documented and many users are not aware of it, leading to very subtle issues such as https://github.com/supercollider/supercollider/pull/6441#issuecomment-2329977621

(I find it a bit strange that an asynchronous method would block depending on its execution context, but we cannot really change it.)

Anyway, this behavior is currently not documented at all!

This applies to several `Buffer` buffer methods and maybe also some other methods. A grep for `forkIfNeeded` may help find all affected methods.

**Comment by muellmusik** [comment]
Can we not fix it? I don’t remember what the intention was but I agree it seems odd.

**Comment by Spacechild1** [comment]
I think we cannot change it without breaking backwards compatibility...

**Comment by telephon** [comment]
Seemed to be a good idea at the time, 15 years ago. The idea was that you could call several of them in order and they would embed in the outer routine.

```supercollider
var buffers;
fork {
   var result;
    buffers.do { |b| b.loadToFloatArray(action: { |x| result = result.add(x) }) }
}
```

this would not be possible as simply with a fork.

Now the problem appears when we write

```supercollider
var buffers;
fork {
   var result, cond;
   cond = CondVar.new;
    buffers.do { |b| 
         var a;
         b.loadToFloatArray(action: { |x|  a = x; cond.signalOne; });
         cond.wait { a.notNil };
         result = result.add(x) 
    }
}
```
It doesn't reach `cond.wait` until it is ready anyway and has called `signalOne`.
I am not sure if there is a general way to resolve this, maybe you have an idea?

**Comment by Spacechild1** [comment]
> The idea was that you could call several of them in order and they would embed in the outer routine.

Yes, that's what I figured. It makes sense given that SC didn't really have a modern async programming model.

(
If the SC Class Library had been written 15 years later, this would probably look as follows:
```
var buffers;
fork {
    // loadToFloatArray would returns a Future
    var result = (buffers.collect { |b| b.loadToFloatArray }).awaitAll;
}
```
)

> I am not sure if there is a general way to resolve this, maybe you have an idea?

As I've written in the other thread, you can just wrap the call to the async function in a Routine:

```
fork {
    var a;
    fork {         
        b.loadToFloatArray(action: { |x|  a = x; cond.signalOne; });
    };
    cond.waitFor(5.0, { a.notNil });
}
```

So IMO it's not a big issue, it just needs documentation.

**Comment by JordanHendersonMusic** [comment]
Just to say the above linked PR found this bug in the tests and reduce execution time from ~100s to ~5s. 
That is a lot of wasted server time (and energy) over the years. 

> So IMO it's not a big issue, it just needs documentation.
@Spacechild1 

Wouldn't you need to know entire call stack and whether any of the methods call `forkIfNeeded`.?

**Comment by Spacechild1** [comment]
> Just to say the above linked PR found this bug in the tests and reduce execution time from ~100s to ~5s.

Just to be clear: the behavior seems to be intentional (see https://github.com/supercollider/supercollider/issues/6448#issuecomment-2332529183), the "but" is only the missing the documentation. Personally, I have been aware of this behavior, but only because I've studied the Class Library code :)

> Wouldn't you need to know entire call stack and whether any of the methods call forkIfNeeded.?

I'm not sure I understand the question. If you want `loadToFloatArray` to always run in its own Routine, just wrap it in `fork`. It doesn't matter if the *calling* code is running in a Routine or not. `loadToFloatArray` itself will run synchronously inside the new Routine created with `fork`.

**Comment by JordanHendersonMusic** [comment]
Ah I was wondering if we have a coloured function situation where using forkIfNeeded in a method would require all callers to exhibit this behaviour, and therefore we need to go around and look at which methods call methods (and so on) that call forkIfNeeded. 

So (if this is true) it's not *just* the methods that directly use forkIfNeeded that needs documentation changes.

**Comment by Spacechild1** [comment]
>  and therefore we need to go around and look at which methods call methods (and so on) that call forkIfNeeded.

I see. Yes, for example, if there was another method somewhere which internally calls `loadToFloatArray`, it *might* be affected as well and require additional documentation.
